### ISSUE TITLE
[Change Request: space-in-parens] Allow Space Before `!` Operator

### ISSUE THREAD

**Issue by Mheetu** [eslint/eslint#13070]
**What rule do you want to change?**
Perhaps `space-in-parens`, but I'm still new to ESLint so there may be a better approach...

I like making unary `not` (`!`) operators stand out and easy to see, so instead of:
```js
if (!someBoolean) {
```
I tend to prefer: 
```js
if ( ! someBoolean) {
``` 
You'll notice I placed a space on both sides of the `!`.

We can acheive the space on the right with `'space-unary-ops': ['error', {'overrides': {'!': true}}],` but `space-in-parens` would prevent us from having one on the left without having to disable it entirely.

**Does this change cause the rule to produce more or fewer warnings?**
Both.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New options that enforce adding a space before unary not `!` after an opening parenthesis.

**Please provide some example code that this change will affect:**
```js
// would error with a new `!` override enabled, regardless of `always` or `never`
if (! noSpaceBeforeUnary) {
```
```js
// would error, if set to `always` with the new `!` override enabled
if (! noSpaceBeforeUnary ) {
```
```js
// would be happy, if set to `always` with the new `!` override enabled
if ( ! noSpaceBeforeUnary ) {
```
```js
// would error, if set to `never` with the new `!` override enabled
if (! noSpaceBeforeUnary) {
```
```js
// would be happy, if set to `never` with the new `!` override enabled
if ( ! noSpaceBeforeUnary) {
```

**What does the rule currently do for this code?**
Currently, you can set it to enforce spaces inside parenthesis or not, there's no override option for cases where the space is before a `!`.

**What will the rule do after it's changed?**
Provide the option to enforce having a space before a `!`, regardless of if spaces are otherwise allowed or disallowed inside parenthesis.

**Are you willing to submit a pull request to implement this change?**
I wish I were able to contribute towards making this a reality, but I am still new to JavaScript... it'll be a while before I'm able to understand how to craft rules for ESLint.

**Comment by mdjermanovic** [comment]
Thanks for the issue! 

Given the existing configuration options in this rule, I think we could add a new [exception](https://eslint.org/docs/rules/space-in-parens#exceptions), rather than an "override" option.

It would, like other "exceptions" in this rule, require the opposite of what's specified as the string `"always"/"never"` option:

```js
/* eslint "space-in-parens": ["error", "always", { "exceptions": ["!"] }] */

if ( ! a ) {} // error

if (! a ) {} // ok
```

```js
/* eslint "space-in-parens": ["error", "never", { "exceptions": ["!"] }] */

if (! a) {} // error

if ( ! a) {} // ok
```

**Comment by kaicataldo** [comment]
The rule applies to both opening and closins parens, so it's a bit strange to me that this one rule would only apply to one side. Can you explain how a user would configure this, particularly if they wanted them to be balanced?

Additionally, can you point us to a popular style guide that has mismatching spaces in parens (i.e. open parens has a space while the closing parens doesn't, or vice versa)? While we would love to accommodate all requests, we have to be judicious with what we include in core. We have a very small team and we need to be responsible maintainers and ensure that we're not adding too much to core at this point. If this isn't recommended by a popular style guide, I'm :-1: for this proposal because it feels like too niche a preference for a core rule.

I would recommend using [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer) to create a custom rule.

**Comment by Mheetu** [comment]
I've honestly never searched for it in a JS style guide... It's a style I've picked up working with PHP that I found to make reading unary not's a lot easier while still reducing spaces where they provide no significant value.

That said, support for this has been added in [WebStorm](https://youtrack.jetbrains.com/issue/WEB-7316). WordPress has a [similar JS style](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#examples-of-good-spacing), but they do have balanced parens so this doesn't affect them.
